### PR TITLE
Add MFSDP v2 expert sharding strategy

### LINKED ISSUE #6941: mFSDP v2 should support heterogeneous FSDP configurations for different parameter groups.
mFSDP v2 should support **heterogeneous FSDP configurations** for different parameter groups.

For users training on NVL8 systems, sharding expert model weights and gradients can introduce significant inter-node communication. When EP spans multiple nodes, this FSDP traffic can contend with inter-node EP communication, slowing it down to the point where the EP communication can no longer be effectively hidden by computation.

A potential compromise would be to apply **ZeRO-2/3-style sharding only to dense parameters within the intra-node NVL domain**, while using only **ZeRO-1-level sharding across the full world size for expert parameters**.

Concretely, this would require enabling HFSDP with `inner_size = 8` for dense parameters, while using a different FSDP configuration for expert parameters rather than applying the same HFSDP setup to both. In other words, mFSDP v2 would need to support different sharding strategies for dense and expert parameters within the same model.

**This feature could help achieve more meaningful performance gains for NVL8 training scenarios. Without it, it would be difficult to determine whether mFSDP or increasing the PP degree would have a smaller impact on overall training performance.** 

@wujingyue, cc @shjwudp  @Autumn1998

### PR DESCRIPTION
**Pull request by Autumn1998**
## What does this PR do?

  Allow MFSDP v2 to use different data-parallel sharding strategies for dense and MoE expert parameters.

| Parameter set | Strategy |
| --- | --- |
| Dense parameters | `data_parallel_sharding_strategy` |
| Expert parameters | `expert_data_parallel_sharding_strategy` |

  The new expert strategy supports all existing MFSDP layouts: no_shard, optim (ZeRO-1), optim_grads (ZeRO-2), and optim_grads_params (ZeRO-3). When expert_data_parallel_sharding_strategy
  is unset, it falls back to data_parallel_sharding_strategy, preserving the existing behavior.

  The implementation translates the dense and expert strategies into separate parameter, gradient, and optimizer placements. 

  ## Validation

  - Unit test covering all combinations of dense and expert sharding strategies and validating loss parity against a `no_shard` reference.
  - Qwen3-MoE-30B-Proxy training from scratch with mock data on 8×H100 GPUs. Loss curves comparing 3D parallelism with MFSDP v2 (`dense`: ZeRO-3, `expert`: ZeRO-1) are available in this
  [W&B report](https://wandb.ai/nvidia/tongliu-qwen3-moe-proxy-h100-fsdpv2/reports/Untitled-Report--VmlldzoxNzgwMDQ2Ng).

### CODE REVIEW

**Review by wujingyue — COMMENTED** [review]
Hey @rapatel, is this aligned with https://github.com/NVIDIA/Megatron-LM/pull/6594? I'm OK with approving both -- I'm just making sure the two agree on adapter-level flags.

**Comment by rapatel** [comment]
> Hey @rapatel, is this aligned with #6594? I'm OK with approving both -- I'm just making sure the two agree on adapter-level flags.

@wujingyue @Autumn1998 These do overlap. #6594 adds the same capability for MFSDP v1 as `data_parallel_sharding_strategy_experts`, which would leave us with two config fields meaning the same thing. I've renamed mine to `expert_data_parallel_sharding_strategy` to match this PR and the `expert_*` convention used elsewhere in MCore (`expert_model_parallel_size`, `expert_tensor_parallel_size`), so we now agree on the field name.

One important thing to note: my PR #6594 adds a guard in V2's config validation that raises if `expert_data_parallel_sharding_strategy` is set, since v1 is the only implementation today. That guard has to be removed by whichever of these lands second. I added it [here](https://github.com/NVIDIA/Megatron-LM/pull/6594/changes#diff-8d3ed77aead9b21b07dfd6c4df4a3b0db421c1f29cf1d1ff52fcc72f8e2eccbfR686-R690).

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 1]
```diff
@@ -63,6 +63,19 @@
 logger = logging.getLogger(__name__)
 
 
+def _placements_from_sharding_strategy(strategy: str) -> Placements:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py]:
```
+class _AxisPlacements(NamedTuple):
+    """How one mesh axis places each of MFSDP's three buffers."""
+
+    parameter: Placement
+    gradient: Placement
+    optimizer: Placement
+
+
+_DATA_PARALLEL_PLACEMENTS = {
+    "no_shard": _AxisPlacements(Replicate(), Partial("avg"), Replicate()),  # DDP
+    "optim": _AxisPlacements(Replicate(), Partial("avg"), Shard(0)),  # ZeRO-1
+    "optim_grads": _AxisPlacements(Replicate(), Shard(0), Shard(0)),  # ZeRO-2
+    "optim_grads_params": _AxisPlacements(Shard(0), Shard(0), Shard(0)),  # ZeRO-3
+}
```

Consider the above. NamedTuple so we don't misremember the integer index. Dict so the conversion is listed as a table -- that said, I'm OK with match-case.

File: megatron/core/distributed/distributed_data_parallel_config.py [hunk 2]
```diff
@@ -104,6 +104,10 @@ class DistributedDataParallelConfig:
     """Sharding strategy for FSDP. Valid values are 'no_shard', 'optim',
       'optim_grads', 'optim_grads_params'."""
 
+    expert_data_parallel_sharding_strategy: Optional[str] = None
```
Comment (wujingyue) [inline megatron/core/distributed/distributed_data_parallel_config.py]:
```suggestion
    expert_data_parallel_sharding_strategy: str | None = None
```

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py [hunk 3]
```diff
@@ -233,5 +267,6 @@ def test_ep_fsdp_matches_fullbatch_reference(distributed_setup):
     )
 
     # Destroy the groups this test created; leave the default (world) group for later tests.
+    dist.barrier()
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: Why?

Comment (Autumn1998) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: I previously observed a hang that appeared to be caused by some ranks reaching cleanup early and destroying the process groups while other ranks were still using them. However, I reran the test several times and could not reproduce it, so I removed the barrier.

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py [hunk 4]
```diff
@@ -37,9 +38,27 @@
 from megatron.core.transformer.transformer_config import TransformerConfig
 from megatron.core.transformer.transformer_layer import MoETransformerLayer
 
-_FLAT_SHARD = Placements(
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:40]: Nit: why did you choose to replace this with a **no-parameter** factory?

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py [hunk 5]
```diff
@@ -221,9 +252,12 @@ def test_ep_fsdp_matches_fullbatch_reference(distributed_setup):
     target = torch.randn(global_batch, seq, vocab, device=device)
     shard = slice(rank * b_local, (rank + 1) * b_local)
 
-    reference_losses = _train(reference, ids, pos, mask, target)
+    reference_optimizer = torch.optim.SGD(reference.parameters(), lr=0.02, foreach=False)
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py]: With fully_shard_optimizer, we can use more common optimizers like Adam. SGD(foreach=False) was used to allow mismatched data/grad dtypes before fully_shard_optimizer existed.

**Review by wujingyue — APPROVED** [review]

File: tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py (hunk 4 above)
Comment (Autumn1998) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:40]: I noticed that most of the other tests use functions for this. Was there a particular reason to keep `_FLAT_SHARD` as a shared constant?

Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_expert_parallel.py:40]: LGTM either way. I was just curious if you had a particular reason. A shared constant is probably better because it's more restrictive and we don't need a function (the YAGNI principle).

**Review by jaredcasper — APPROVED** [review]

**Review by kunlunl — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]
